File: desktop/internal/i18n/i18n.go
// Package i18n provides minimal, native-layer localization for the desktop
// shell. Window titles and the process app name are set in the OS layer (Wails
// Options.Name / WebviewWindowOptions.Title) where the SPA's i18n is not
// reachable, so the desktop binary needs its own tiny dictionary.
//
// Scope is intentionally small: brand name plus the local/remote/manage labels
// used to assemble window titles. The language is resolved once from the OS
// locale at startup (see DetectLang); changing the system language requires a
// restart to take effect, which is acceptable for native chrome.
package i18n

import "strings"

// Dictionary keys. Kept as exported constants so callers don't pass bare
// strings that can drift from the map below.
const (
	// KeyBrand is the short brand used as the window-title prefix
	// ("牛牛" / "Niuniu"), e.g. "牛牛 · 本地".
	KeyBrand = "BRAND"
	// KeyAppName is the full product/app name ("牛牛桌面版" / "Niuniu Desktop")
	// used for the process app name (Options.Name → macOS Dock / Windows taskbar
	// group) and the tray tooltip — NOT the window title.
	KeyAppName = "APP_NAME"
	KeyLocal   = "LOCAL"
	KeyRemote  = "REMOTE"
	KeyManage  = "MANAGE"
	// KeyAI is the AI-aggregation window label ("AI 直达" / "AI Hub").
	KeyAI = "AI"
	// KeyRunners is the global local-runner manager label
	// ("执行器管理" / "Runner Manage").
	KeyRunners = "RUNNERS"
	// KeyBooting is the local boot-splash heading verb ("启动" / "Starting"),
	// composed with AppName into "启动 牛牛桌面版…" / "Starting Niuniu Desktop…"
	// (see LocalBootHeading).
	KeyBooting = "BOOTING"
	// KeyInitLocalService is the local boot-splash sub-status shown while the
	// embedded server spawns ("正在初始化本地服务" / "Initializing local service").
	KeyInitLocalService = "INIT_LOCAL_SERVICE"
)

// defaultLang is used both as the dictionary fallback and as the result of
// DetectLang when the OS locale can't be resolved or isn't recognized.
const defaultLang = "en"

// sep joins the brand-prefixed title segments: "{BRAND} · {…}". The middle dot
// (U+00B7) with surrounding spaces matches the spec examples
// ("牛牛 · 本地" / "Niuniu · Local").
const sep = " · "

// dict maps lang → key → translation. Add new languages here; DetectLang only
// distinguishes zh from the en default today, but T tolerates any lang present.
var dict = map[string]map[string]string{
	"en": {
		KeyBrand:            "Niuniu",
		KeyAppName:          "Niuniu Desktop",
		KeyLocal:            "Local",
		KeyRemote:           "Remote",
		KeyManage:           "Manage Connections",
		KeyAI:               "AI Hub",
		KeyRunners:          "Runner Manage",
		KeyBooting:          "Starting",
		KeyInitLocalService: "Initializing local service",
	},
	"zh": {
		KeyBrand:            "牛牛",
		KeyAppName:          "牛牛桌面版",
		KeyLocal:            "本地",
		KeyRemote:           "远端",
		KeyManage:           "管理连接",
		KeyAI:               "AI 直达",
		KeyRunners:          "执行器管理",
		KeyBooting:          "启动",
		KeyInitLocalService: "正在初始化本地服务",
	},
}

// T returns the translation for key in lang, falling back to the default
// language (and finally the raw key) when a lang or key is missing.
func T(lang, key string) string {
	if m, ok := dict[lang]; ok {
		if v, ok := m[key]; ok {
			return v
		}
	}
	if v, ok := dict[defaultLang][key]; ok {
		return v
	}
	return key
}

// Normalize maps an arbitrary OS locale string (e.g. "zh-CN", "zh_Hans_CN",
// "en_US.UTF-8") to one of the supported language codes. Anything that begins
// with "zh" is Chinese; everything else (including empty) is the default.
func Normalize(locale string) string {
	l := strings.ToLower(strings.TrimSpace(locale))
	if strings.HasPrefix(l, "zh") {
		return "zh"
	}
	return defaultLang
}

// DetectLang resolves the OS locale (platform-specific, see detect_*.go) and
// normalizes it. Callers should cache the result once at startup.
func DetectLang() string {
	return Normalize(osLocale())
}

// Brand returns the short localized brand ("牛牛" / "Niuniu") used as the
// window-title prefix.
func Brand(lang string) string {
	return T(lang, KeyBrand)
}

// AppName returns the full product/app name ("牛牛桌面版" / "Niuniu Desktop")
// for the process app name (Options.Name → macOS Dock / Windows taskbar group)
// and the tray tooltip.
func AppName(lang string) string {
	return T(lang, KeyAppName)
}

// LocalTitle is the title for the local main window: "{BRAND} · {LOCAL}".
func LocalTitle(lang string) string {
	return T(lang, KeyBrand) + sep + T(lang, KeyLocal)
}

// LocalBootHeading is the local boot-splash heading: "{BOOTING} {APP_NAME}…"
// ("启动 牛牛桌面版…" / "Starting Niuniu Desktop…"), shown while the embedded
// server spawns.
func LocalBootHeading(lang string) string {
	return T(lang, KeyBooting) + " " + AppName(lang) + "…"
}

// RemoteTitle is the title for a remote connection window:
// "{BRAND} · {REMOTE} · {name} ({hostPort})".
func RemoteTitle(lang, name, hostPort string) string {
	return T(lang, KeyBrand) + sep + T(lang, KeyRemote) + sep + name + " (" + hostPort + ")"
}

// ManageTitle is the title for the connection-manager (picker) window:
// "{BRAND} · {MANAGE}".
func ManageTitle(lang string) string {
	return T(lang, KeyBrand) + sep + T(lang, KeyManage)
}

// AITitle is the title for the AI-aggregation hub window: "{BRAND} · {AI}".
func AITitle(lang string) string {
	return T(lang, KeyBrand) + sep + T(lang, KeyAI)
}

// RunnersTitle is the title for the global local-runner manager window:
// "{BRAND} - {RUNNERS}" ("牛牛 - 执行器管理" / "Niuniu - Runner Manage"). Uses a
// hyphen separator (not the middle dot the other windows use) per the spec'd
// label.
func RunnersTitle(lang string) string {
	return T(lang, KeyBrand) + " - " + T(lang, KeyRunners)
}

// AIServiceTitle is the title for one AI service's webview window:
// "{BRAND} · {AI} · {name}".
func AIServiceTitle(lang, name string) string {
	return T(lang, KeyBrand) + sep + T(lang, KeyAI) + sep + name
}

File: desktop-v2/src/i18n.rs
//! 极简原生层本地化：窗口标题 / 托盘文案在 OS 层由本字典给出（SPA 自己的 i18n
//! 在这里不可达，与 Wails 版 internal/i18n 同构）。语言在启动时解析一次。

/// 解析系统 locale（如 "zh-CN" / "en_US"），归一化为 "zh" 或 "en"。
pub fn detect_lang() -> &'static str {
    match sys_locale::get_locale() {
        Some(l) if l.to_lowercase().starts_with("zh") => "zh",
        _ => "en",
    }
}

fn t(lang: &str, key: &str) -> &'static str {
    let zh: bool = lang == "zh";
    match key {
        "brand" => {
            if zh {
                "牛牛"
            } else {
                "Niuniu"
            }
        }
        "app_name" => {
            if zh {
                "牛牛桌面版"
            } else {
                "Niuniu Desktop"
            }
        }
        "local" => {
            if zh {
                "本地"
            } else {
                "Local"
            }
        }
        "remote" => {
            if zh {
                "远端"
            } else {
                "Remote"
            }
        }
        "manage" => {
            if zh {
                "管理连接"
            } else {
                "Manage Connections"
            }
        }
        "ai" => {
            if zh {
                "AI 直达"
            } else {
                "AI Hub"
            }
        }
        "runners" => {
            if zh {
                "执行器管理"
            } else {
                "Runner Manage"
            }
        }
        "booting" => {
            if zh {
                "启动"
            } else {
                "Starting"
            }
        }
        "init_local_service" => {
            if zh {
                "正在初始化本地服务"
            } else {
                "Initializing local service"
            }
        }
        _ => "?",
    }
}

/// 全称产品名（进程 app name / 托盘 tooltip）。
pub fn app_name(lang: &str) -> &'static str {
    t(lang, "app_name")
}

/// 短品牌名（窗口标题前缀）。
pub fn brand(lang: &str) -> &'static str {
    t(lang, "brand")
}

/// 主窗口标题 "{品牌} · {本地}"。
pub fn local_title(lang: &str) -> String {
    format!("{} · {}", brand(lang), t(lang, "local"))
}

/// 主窗口加载过渡页标题 "{启动} {全称}…"。
pub fn local_boot_heading(lang: &str) -> String {
    format!("{} {}…", t(lang, "booting"), app_name(lang))
}

/// picker（连接管理）窗口标题。
pub fn manage_title(lang: &str) -> String {
    format!("{} · {}", brand(lang), t(lang, "manage"))
}

/// AI 直达窗口标题。
pub fn ai_title(lang: &str) -> String {
    format!("{} · {}", brand(lang), t(lang, "ai"))
}

/// 执行器管理窗口标题。
pub fn runners_title(lang: &str) -> String {
    format!("{} - {}", brand(lang), t(lang, "runners"))
}

/// 托盘菜单用的「执行器管理」裸标签（不带品牌前缀，对应 v1 app.go 的
/// i18n.T(lang, KeyRunners)）。
pub fn t_runners(lang: &str) -> &'static str {
    if lang == "zh" { "执行器管理" } else { "Runner Manage" }
}

/// 远端连接窗口标题。
pub fn remote_title(lang: &str, name: &str, host_port: &str) -> String {
    format!("{} · {} · {} ({})", brand(lang), t(lang, "remote"), name, host_port)
}
